src/python/pants/core/util_rules/archive.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from dataclasses import dataclass
+from typing import Optional, Tuple
+
+from pants.engine.fs import Digest, DirectoryWithPrefixToStrip, Snapshot
+from pants.engine.process import Process, ProcessResult
+from pants.engine.rules import RootRule, rule
+from pants.engine.selectors import Get
+
+
+@dataclass(frozen=True)
+class MaybeExtractable:
+    """A newtype to work around rule graph resolution issues.
+
+    Possibly https://github.com/pantsbuild/pants/issues/9320. Either way, we should fix the
+    underlying issue and get rid of this type.
+    """
+
+    digest: Digest
+
+
+@dataclass(frozen=True)
+class ExtractedDigest:
+    """The result of extracting an archive."""
+
+    digest: Digest
+
+
+def get_extraction_cmd(archive_path: str, output_dir: str) -> Optional[Tuple[str, ...]]:
+    """Returns a shell command to run to extract the archive to the output directory."""
+    # Note that we assume that mkdir, unzip, and tar exist in the executing environment.
+    if archive_path.endswith(".zip"):
+        return ("unzip", "-q", archive_path, "-d", output_dir)
+    elif archive_path.endswith((".tar", ".tar.gz", ".tgz", ".tar.bz2", ".tbz2", ".tar.xz", ".txz")):
+        return ("mkdir", output_dir, "&&", "tar", "xf", archive_path, "-C", output_dir)
+    return None
+
+
+@rule
+async def maybe_extract(extractable: MaybeExtractable) -> ExtractedDigest:
+    """If digest contains a single archive file, extract it, otherwise return the input digest."""
+    digest = extractable.digest
+    snapshot = await Get[Snapshot](Digest, digest)
+    if len(snapshot.files) == 1:
```

Comment:
Conditional extraction doesn't feel like a good default, because it's a bit magical. Would be good to expose an API the asserts that something is an archive as-well/by-default.

It might even be worth having BinaryUtil internally implement the "maybe extract" logic?

Reply:
Hmm, conditional extraction requires having the Snapshot, which this code needs anyway but the binary util code wouldn't.  How about I add a "passthrough" field on MaybeExtractable (and change its name) that indicates what the behavior should be if this isn't an archive?

Reply:
Actually on further thought, I'm not sure this is magical. I think of an archive as "a mapping of file names to file content" and a regular old directory is such a thing as well. It's like a no-op archive... Maybe I'll leave this for a followup?

Reply:
OTOH I hear you that this might hide errors. Let me try and come up with something.

Reply:
Hmm, it starts to get complicated. I'd prefer to revisit if/when it's a problem, and after the external tool code is in, if that's OK.